server/lib/chat/chat-api.ts

```diff
@@ -0,0 +1,206 @@
+import { MULTI_CHANNEL } from '../../../common/flags'
+import { featureEnabled } from '../flags/feature-enabled'
+import { httpApi, HttpApi } from '../http/http-api'
+import ensureLoggedIn from '../session/ensure-logged-in'
+import createThrottle from '../throttle/create-throttle'
+import throttleMiddleware from '../throttle/middleware'
+import { validateRequest } from '../validation/joi-validator'
+import ChatService, { ChatServiceError, ChatServiceErrorCode } from './chat-service'
+
+const joinThrottle = createThrottle('chatjoin', {
+  rate: 3,
+  burst: 10,
+  window: 60000,
+})
+
+const leaveThrottle = createThrottle('chatleave', {
+  rate: 10,
+  burst: 20,
+  window: 60000,
+})
+
+const sendThrottle = createThrottle('chatsend', {
+  rate: 30,
+  burst: 90,
+  window: 60000,
+})
+
+const retrievalThrottle = createThrottle('chatretrieval', {
+  rate: 30,
+  burst: 120,
+  window: 60000,
+})
+
+function isChatServiceError(error: Error): error is ChatServiceError {
+  return error.hasOwnProperty('code')
+}
+
+function convertChatServiceError(err: Error) {
+  if (!isChatServiceError(err)) {
+    throw err
+  }
+
+  switch (err.code) {
+    case ChatServiceErrorCode.UserOffline:
+      throw new httpErrors.NotFound(err.message)
+    case ChatServiceErrorCode.InvalidJoinAction:
+    case ChatServiceErrorCode.InvalidLeaveAction:
+    case ChatServiceErrorCode.InvalidSendAction:
+    case ChatServiceErrorCode.InvalidGetHistoryAction:
+    case ChatServiceErrorCode.InvalidGetUsersAction:
+      throw new httpErrors.BadRequest(err.message)
+    case ChatServiceErrorCode.LeaveShieldBattery:
+      throw new httpErrors.Forbidden(err.message)
+    default:
+      assertUnreachable(err.code)
+  }
+}
+
+@httpApi()
+export class ChatApi extends HttpApi {
+  constructor() {
+    super('/chat')
+    container.resolve(ChatService)
+  }
+
+  applyRoutes(router: Router): void {
+    router
+      .use(featureEnabled(MULTI_CHANNEL), ensureLoggedIn)
+      .post(
+        '/:channelName',
+        throttleMiddleware(joinThrottle, ctx => String(ctx.session!.userId)),
+        joinChannel,
+      )
+      .delete(
+        '/:channelName',
+        throttleMiddleware(leaveThrottle, ctx => String(ctx.session!.userId)),
+        leaveChannel,
+      )
+      .patch(
+        '/:channelName',
+        throttleMiddleware(sendThrottle, ctx => String(ctx.session!.userId)),
+        sendChatMessage,
+      )
+      .get(
+        '/:channelName/chat',
+        throttleMiddleware(retrievalThrottle, ctx => String(ctx.session!.userId)),
+        getChannelHistory,
+      )
+      .get(
+        '/:channelName/users',
+        throttleMiddleware(retrievalThrottle, ctx => String(ctx.session!.userId)),
+        getChannelUsers,
+      )
+  }
+}
+
+function joinChannel(ctx: RouterContext) {
+  const {
+    params: { channelName },
+  } = validateRequest(ctx, {
+    params: Joi.object<{ channelName: string }>({
```

Comment:
Would probably be good to pull both the interface and the validation out somewhere common, since this is the same across all these APIs